File: app/application/socketio.py
from flask import request
from flask_socketio import emit, join_room, leave_room, close_room
from app import socketio

socketio_cbs = {}

# clients can subscribe to a room.  The server can broadcast an event into that room only.


@socketio.event
def subscribe_to_room(message):
    join_room(message['room'])


@socketio.event
def leave_room(message):
    leave_room(message['room'])


@socketio.on
def close_room(message):
    close_room(message['room'])


@socketio.event
def send_to_server(msg):
    if msg['type'] in socketio_cbs:
        if 'value' in msg['data'] and msg['data']['value'] == 'on': # bugfix socketio?  Sometimes, a True at the sender becomes 'on' at the receiver
            msg['data']['value'] = True
        for cb in socketio_cbs[msg['type']]:
            cb(msg, request.sid)


@socketio.on('disconnect')
def disconnect_socket():
    if 'disconnect' in socketio_cbs:
        for cb in socketio_cbs['disconnect']:
            cb(None, request.sid)


@socketio.event
def connect():
    pass


@socketio.event
def its_me(data):
    emit('send_to_client', {'type': 'its-me-received', 'data': True}, room=request.sid)


def subscribe_on_type(type, cb):
    if type in socketio_cbs:
        socketio_cbs[type].append(cb)
    else:
        socketio_cbs[type] = [cb]


def send_to_room(msg, room):
    emit('send_to_client', msg, room=room)


def broadcast_message(msg):
    emit('send_to_client', msg, broadcast=True, namespace='/')


def send_to_client(client_sid, type, msg):
    emit(type, msg, room=client_sid)

File: app/static/js/socketio.js
class Socketio {
    receive_cbs = {}
    constructor() {
        this.socket = io();
        this.subscribe_on_receive("its-me-received", this.handle_its_me_received)
    }

    start(on_connect_cb, opaque) {
        this.socket.on('send_to_client', function (msg, cb) {
            if (msg.type in this.receive_cbs) {
                this.receive_cbs[msg.type](msg.type, msg.data);
            }
            if (cb)
                cb();
        }.bind(this));

        this.socket.on('connect', function () {
            if (on_connect_cb) {
                on_connect_cb(opaque);
            }
        }.bind(this));

    }

    send_to_server(type, data) {
        this.socket.emit('send_to_server', {type: type, data: data});
        return false;
    }

    subscribe_on_receive(type, receive_cb) {
        this.receive_cbs[type] = receive_cb;
    }

    handle_its_me_received(type, data) {
        console.log("its-me-received: " + data);
    }

    subscribe_to_room(room_code) {
        this.socket.emit('subscribe_to_room', {room: room_code});
    }
}

var socketio
$(function () { // same as $(document).ready()
    socketio = new Socketio();
});

